pilot/pkg/serviceregistry/kube/controller/ambient/sidecar_interop.go

```diff
@@ -0,0 +1,173 @@
+}
+
+func (w workloadEDS) ResourceName() string {
+	return w.WorkloadKey
+}
+
+// RegisterEdsShim handles triggering xDS events when Envoy EDS needs to change.
+// Most of ambient index works to build `workloadapi` types - Workload, Service, etc.
+// Envoy uses a different API, with different relationships between types.
+// To ensure Envoy are updated properly on changes, we compute this information.
+// Currently, this is only used to trigger events.
+// Ideally, the information we are using in Envoy and the event trigger are using the same data directly.
+func RegisterEdsShim(
+	xdsUpdater model.XDSUpdater,
+	Workloads krt.Collection[model.WorkloadInfo],
+	WorkloadsByServiceKey krt.Index[string, model.WorkloadInfo],
+	Services krt.Collection[model.ServiceInfo],
+	ServicesByAddress krt.Index[networkAddress, model.ServiceInfo],
+) {
+	// When sending information to a workload, we need two bits of information:
+	// * Does it support tunnel? If so, we will need to use HBONE
+	// * Does it have a waypoint? if so, we will need to send to the waypoint
+	// Record both of these.
+	// Note: currently, EDS uses the waypoint VIP for workload waypoints. This is probably something that will change.
+	WorkloadEds := krt.NewCollection(
+		Workloads,
+		func(ctx krt.HandlerContext, wl model.WorkloadInfo) *workloadEDS {
+			if wl.Waypoint == nil && wl.TunnelProtocol == workloadapi.TunnelProtocol_NONE {
+				return nil
+			}
+			res := &workloadEDS{
+				WorkloadKey:      wl.ResourceName(),
+				TunnelProtocol:   wl.TunnelProtocol,
+				ServiceHostnames: slices.Sort(maps.Keys(wl.Services)),
+			}
+			if wl.Waypoint != nil {
+				if wl.Waypoint.GetHostname() == nil {
+					// TODO!!!!!
+					panic("todo")
+				}
+				hn := wl.Waypoint.GetHostname()
+				waypointServiceKey := hn.Namespace + "/" + hn.Hostname
+				workloads := krt.Fetch(ctx, Workloads, krt.FilterIndex(WorkloadsByServiceKey, waypointServiceKey))
+				res.WaypointInstance = slices.Map(workloads, func(e model.WorkloadInfo) *workloadapi.Workload {
+					return e.Workload
+				})
+			}
+			return res
+		},
+		krt.WithName("WorkloadEds"))
+	ServiceEds := krt.NewCollection(
+		Services,
+		func(ctx krt.HandlerContext, svc model.ServiceInfo) *serviceEDS {
+			if svc.Service.Waypoint == nil {
+				return nil
+			}
+			if svc.Service.Waypoint.GetHostname() == nil {
+				// TODO!!!!!
+				panic("todo")
+			}
+			hn := svc.Service.Waypoint.GetHostname()
+			waypointServiceKey := hn.Namespace + "/" + hn.Hostname
+			workloads := krt.Fetch(ctx, Workloads, krt.FilterIndex(WorkloadsByServiceKey, waypointServiceKey))
+			return &serviceEDS{
+				ServiceKey: svc.ResourceName(),
+				WaypointInstance: slices.Map(workloads, func(e model.WorkloadInfo) *workloadapi.Workload {
+					return e.Workload
+				}),
+			}
+		},
+		krt.WithName("ServiceEds"))
+	WorkloadEds.RegisterBatch(
+		PushAllXds(xdsUpdater, func(i workloadEDS, updates sets.Set[model.ConfigKey]) {
+			for _, svc := range i.ServiceHostnames {
```

Comment:
Probably just my ignorance, but I'm confused about the relationship between workload addressing and service hostnames. Is this for headless service?